src/persist/src/workload.rs

```diff
@@ -0,0 +1,274 @@
+            RECORD_SIZE_BYTES_KEY,
+            record_size_bytes,
+            BATCH_MAX_COUNT_KEY,
+            batch_max_count,
+        );
+        DataGenerator::new(record_count, record_size_bytes, batch_max_count)
+    }
+}
+
+impl DataGenerator {
+    /// Returns a new [DataGenerator].
+    pub fn new(record_count: usize, record_size_bytes: usize, batch_max_count: usize) -> Self {
+        // NB: Strict greater so we have at least one byte for key.
+        assert!(record_size_bytes > TS_DIFF_GOODPUT_SIZE);
+        let key_len = record_size_bytes - TS_DIFF_GOODPUT_SIZE;
+        assert!(batch_max_count > 0);
+        DataGenerator {
+            record_count,
+            record_size_bytes,
+            batch_max_count,
+            key_buf: Vec::with_capacity(key_len),
+            val_buf: Vec::new(),
+        }
+    }
+
+    /// Returns the number of "goodput" bytes represented by the entire dataset
+    /// produced by this generator.
+    pub fn goodput_bytes(&self) -> u64 {
+        u64::cast_from(self.record_count * self.record_size_bytes)
+    }
+
+    /// Returns a more easily human readable version of [Self::goodput_bytes].
+    pub fn goodput_pretty(&self) -> String {
+        let goodput_bytes = self.goodput_bytes();
+        const KIB: u64 = 1024;
+        const MIB: u64 = 1024 * KIB;
+        const GIB: u64 = 1024 * MIB;
+        if goodput_bytes >= 10 * GIB {
+            format!("{}GiB", goodput_bytes / GIB)
+        } else if goodput_bytes >= 10 * MIB {
+            format!("{}MiB", goodput_bytes / MIB)
+        } else if goodput_bytes >= 10 * KIB {
+            format!("{}KiB", goodput_bytes / KIB)
+        } else {
+            format!("{}B", goodput_bytes)
+        }
+    }
+
+    /// Generates the requested batch of records.
+    pub fn gen_batch(&mut self, batch_idx: usize) -> Option<ColumnarRecords> {
+        let batch_start = self.batch_max_count * batch_idx;
+        let batch_end = cmp::min(batch_start + self.batch_max_count, self.record_count);
+        eprintln!(
+            "WIP cfg={:?} batch_idx={} batch_start={} batch_end={}",
+            self, batch_idx, batch_start, batch_end
+        );
+        if batch_start >= batch_end {
+            return None;
+        }
+        let mut batch = ColumnarRecordsBuilder::default();
+        batch.reserve(
+            batch_end - batch_start,
+            self.record_size_bytes - TS_DIFF_GOODPUT_SIZE,
+            0,
+        );
+        for record_idx in batch_start..batch_end {
+            batch.push(self.gen_record(record_idx));
+        }
+        Some(batch.finish())
+    }
+
+    fn gen_record(&mut self, record_idx: usize) -> ((&[u8], &[u8]), u64, isize) {
+        assert!(record_idx < self.record_count);
+        assert!(self.record_size_bytes > TS_DIFF_GOODPUT_SIZE);
+
+        self.key_buf.clear();
+        let key_len = self.record_size_bytes - TS_DIFF_GOODPUT_SIZE;
+        write!(&mut self.key_buf, "{:01$}", record_idx, key_len)
```

Comment:
I glossed over this earlier but can you add a comment describing what this will write out? I looked at the std::fmt docs but it wasn't obvious. similarly, its not 100% clear to me whether this output is in sorted order with respect to the memcmp ordering that vec<u8>s follow

Reply:
sorry i forgot this in the first pass!

Reply:
yeah will do, the summary is printing an integer padded to the given len with 0s. what's the relevance of "sorted order with respect to the memcmp ordering that vecs follow"?

Reply:
well if we are populating the data with 0, 1, 2, 3.. record_count represented as `Vec<u8>` that data isn't actually sorted (because `Vec<u8>`'s are sorted by comparing bytes directly, so in a little endian system e.g. 4096 is sorted before 3). This matters because one of the write benchmarks compares a sorted batch with an unsorted batch, but it shuffles the records it gets from the generator, when it perhaps needs to sort them instead?